eclipse-che/che-server
# Respect the ScmComunicationException on token validation step

### What does this PR do?
* Throw the `ScmCommunicationException` on token validation step instead of returning invalid result.
* Propagate the `ScmCommunicationException` to the dashboard as an `ApiException`.
* Throw `ScmUnauthorizedException` on each scm api request with invalid token.
* Log error instead of throwing `RuntimeException` on workspace provision steps.

### Screenshot/screencast of this PR



### What issues does this PR fix or reference?

https://github.com/eclipse-che/che/issues/23076

### How to test this PR?

1. Deploy Che from the PR image: `quay.io/ivinokur/che-server:che-23076` The image is built with a modification: to simulate the unreachable SCM server I changed the api endpoint url in the GitHub api client:
https://github.com/eclipse-che/che-server/blob/e04d787ed847f06faf9d18c56f0774de3cfa8b7c/wsmaster/che-core-api-factory-github-common/src/main/java/org/eclipse/che/api/factory/server/github/GithubApiClient.java#L57
2. Configure Che to use dashboard image from the related pull request: `quay.io/eclipse/che-dashboard:pr-1179`
3. Configure [GitHub oauth](https://eclipse.dev/che/docs/stable/administration-guide/configuring-oauth-2-for-github/)
4. Start a workspace from a GitHub repository with devfile.

See: a warning notification appears during workspace startup:
![Screenshot 2024-08-30 at 13 59 48](https://github.com/user-attachments/assets/57dd26d4-28ed-4cfc-b658-cf376b812ec1)
The workspace starts successfully. 


### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Release Notes

### Reviewers

Reviewers, please comment how you tested the PR when approving it.
